staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```
@@ -484,3 +490,192 @@ func TestCacheWatcherStoppedInAnotherGoroutine(t *testing.T) {
+		allowWatchBookmarks bool
+		expectedBookmarks   bool
+	}{
+		{
+			watchCacheEnabled:   true,
+			allowWatchBookmarks: true,
+			expectedBookmarks:   true,
+		},
+		{
+			watchCacheEnabled:   true,
+			allowWatchBookmarks: false,
+			expectedBookmarks:   false,
+		},
+		{
+			watchCacheEnabled:   false,
+			allowWatchBookmarks: true,
+			expectedBookmarks:   false,
+		},
+		{
+			watchCacheEnabled:   false,
+			allowWatchBookmarks: false,
+			expectedBookmarks:   false,
+		},
+	}
+
+	for _, tc := range testCases {
+		testCacherSendBookmarkEvents(t, tc.watchCacheEnabled, tc.allowWatchBookmarks, tc.expectedBookmarks)
+	}
+}
+
+func TestDispatchingBookmarkEventsWithConcurrentStop(t *testing.T) {
+	defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.WatchBookmark, true)()
+	backingStorage := &dummyStorage{}
+	cacher, _ := newTestCacher(backingStorage, 1000)
+	defer cacher.Stop()
+
+	// Wait until cacher is initialized.
+	cacher.ready.wait()
+
+	// Ensure there is some budget for slowing down processing.
+	cacher.dispatchTimeoutBudget.returnUnused(100 * time.Millisecond)
+
+	resourceVersion := uint64(1000)
+	err := cacher.watchCache.Add(&examplev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            fmt.Sprintf("pod-0"),
+			Namespace:       "ns",
+			ResourceVersion: fmt.Sprintf("%v", resourceVersion),
+		}})
+	if err != nil {
+		t.Fatalf("failed to add a pod: %v", err)
+	}
+
+	for i := 0; i < 1000; i++ {
+		w, err := cacher.Watch(context.TODO(), "pods/ns", "999", storage.Everything)
+		if err != nil {
+			t.Fatalf("Failed to create watch: %v", err)
+		}
+		bookmark := &watchCacheEvent{
+			Type:            watch.Bookmark,
+			ResourceVersion: uint64(i),
+			Object:          cacher.newFunc(),
+		}
+		cacher.versioner.UpdateObject(bookmark.Object, bookmark.ResourceVersion)
+
+		go func() {
+			cacher.dispatchEvent(bookmark)
+		}()
+
+		go func() {
+			w.Stop()
+		}()
+
+		receiveResult := func() <-chan struct{} {
```

nit: I think it would be more natural to write:

done := make(chan struct{}))
go func() {
  ...
}()

and use just done channel below

---

thanks, and should 'Fatal' to fail fast in the following timeout.